dissect/target/plugins/scrape/qfind.py

```diff
@@ -102,68 +136,75 @@ def qfind(
             needle_lookup = tmp
 
+        elif regex:
+            tmp = {}
+            for needle, _ in needle_lookup.items():
+                tmp[re.compile(needle, re.IGNORECASE if ignore_case else re_NOFLAG)] = _
+            needle_lookup = tmp
+
         seen = set()
-        for _, stream, needle, offset in self.target.scrape.find(
-            list(needle_lookup.keys()), progress=progress(self.target)
+        for _, stream, needle, offset, match in self.target.scrape.find(
+            list(needle_lookup.keys()), progress=progress_handler(self.target) if progress else None
         ):
             original_needle, codec = needle_lookup[needle]
             needle_len = len(needle.pattern if isinstance(needle, re.Pattern) else needle)
             before_offset = max(0, offset - window)
             stream.seek(before_offset)
             buf = stream.read((offset - before_offset) + max(window, needle_len))
 
-            header = f"\r[{offset:#08x} @ {original_needle} ({codec})]"
+            if unique:
+                if window > 20:
+                    m = hashlib.sha1()
+                    m.update(buf)
+                    digest = m.digest()
+                else:
+                    digest = buf
 
-            if raw:
-                print(header)
-                palette = [(offset - before_offset, utils.COLOR_NORMAL), (needle_len, utils.COLOR_BG_RED)]
-                utils.hexdump(buf, palette, offset=before_offset)
-            else:
-                codec = "utf-8" if codec == "hex" else codec
-                before_part = recover_string(
-                    buf[: offset - before_offset], codec, reverse=True, ascii=not allow_non_ascii
-                )
-                after_part = recover_string(buf[offset - before_offset :], codec, ascii=not allow_non_ascii)
-                hit = (
-                    before_part
-                    + utils.COLOR_BG_RED
-                    + after_part[:needle_len]
-                    + utils.COLOR_NORMAL
-                    + after_part[needle_len:]
-                )
-
-                if unique:
-                    if hit in seen:
-                        continue
-                    seen.add(hit)
-
-                print(header)
-                print(hit)
-
-
-def progress(target: Target) -> Callable[[Container | Volume, int, int], None]:
+                if digest in seen:
+                    continue
+                seen.add(digest)
+
+            match = match.group() if isinstance(needle, re.Pattern) and match else original_needle
+
+            yield QFindMatchRecord(
+                offset=offset,
+                needle=original_needle,
+                codec=codec,
+                match=match,
+                content=buf.strip(b"\x00") if strip_null_bytes else buf,
+                _target=self.target if self.target._os else None,
+            )
+
+
+def progress_handler(target: Target) -> Callable[[Container | Volume, int, int], None]:
```

Comment:
I'd move this to `tools/qfind.py` and make the `progress` kwarg a `Callable[[Container | Volume, int, int], None]`.